vembrane/__init__.py

```diff
@@ -112,86 +151,134 @@ def get_annotation_keys(header: VariantHeader, ann_key: str) -> List[str]:
 
 @lru_cache(maxsize=32)
 def parse_annotation_entry(entry: str,) -> List[str]:
-    return list(map(str.strip, entry.split("|")))
+    return entry.split("|")
 
 
-class Expression:
-    def __init__(self, expression: str, ann_key: str = "ANN"):
-        self._expression = expression
+UNSET = object()
+
+
+class Environment(dict):
+    def __init__(self, expression: str, ann_key: str, header: VariantHeader):
         self._ann_key = ann_key
         self._has_ann = any(
             hasattr(node, "id") and isinstance(node, ast.Name) and node.id == ann_key
             for node in ast.walk(ast.parse(expression))
         )
-
-    def annotation_key(self):
-        return self._ann_key
+        self._annotation = Annotation(ann_key, header)
+        self._globals = {}
+        # We use self + self.func as a closure.
+        self._globals = globals_whitelist.copy()
+        self._func = eval(f"lambda: {expression}", self, {})
+
+        self._getters = {
+            "CHROM": self._get_chrom,
+            "POS": self._get_pos,
+            "ID": self._get_id,
+            "ALT": self._get_alt,
+            "REF": self._get_ref,
+            "QUAL": self._get_qual,
+            "FILTER": self._get_filter,
+            "INFO": self._get_info,
+            "FORMAT": self._get_format,
+            "SAMPLES": self._get_samples,
+        }
+        self._empty_globals = {name: NA for name in header.info}
+        self._empty_globals.update({name: UNSET for name in self._getters})
+        self.record = None
+        self.idx = -1
 
     def filters_annotations(self):
         return self._has_ann
 
-    def evaluate(
-        self, idx: int, annotation: str, annotation_keys: List[str], env: dict,
-    ) -> bool:
-        env[self._ann_key] = Annotation(
-            idx,
-            dict(
-                map(
-                    lambda v: ANN_TYPER.convert(v[0], v[1]),
-                    zip(annotation_keys, parse_annotation_entry(annotation)),
-                )
-            ),
-        )
-        return eval(self._expression, globals_whitelist, env)
+    def update(self, idx, record):
+        self.idx = idx
+        self.record = record
+        self._globals.update(self._empty_globals)
+
+    def _get_chrom(self):
+        value = self.record.chrom
+        self._globals["CHROM"] = value
+        return value
+
+    def _get_pos(self):
+        value = self.record.pos
+        self._globals["POS"] = value
+        return value
+
+    def _get_id(self):
+        value = self.record.id
+        self._globals["ID"] = value
+        return value
+
+    def _get_ref_alt(self):
+        ref, *alt = chain(self.record.alleles)
+        self._globals["REF"], self._globals["ALT"] = ref, alt
+        return ref, alt
+
+    def _get_ref(self):
+        return self._get_ref_alt()[0]
+
+    def _get_alt(self):
+        return self._get_ref_alt()[1]
+
+    def _get_qual(self):
```

Comment:
I guess it's a float?
```suggestion
    def _get_qual(self) -> float:
```

Reply:
Looks like `Optional[float]`: https://github.com/pysam-developers/pysam/blob/v0.16.0.1/pysam/libcbcf.pyx#L3117

Reply:
FYI, it looks like the other fields are `Optional` also. Which also include `alleles`, meaning `ref, *alt =` would fail in that case -- but then things seem to be strange in any case..